source/common/stats/symbol_table_impl.cc

```diff
@@ -410,27 +409,56 @@ SymbolTable::StoragePtr SymbolTableImpl::encode(absl::string_view name) {
 }
 
 StatNameStorage::StatNameStorage(absl::string_view name, SymbolTable& table)
-    : bytes_(table.encode(name)) {}
+    : StatNameStorageBase(table.encode(name)) {}
 
 StatNameStorage::StatNameStorage(StatName src, SymbolTable& table) {
   const uint64_t size = src.size();
   MemBlockBuilder<uint8_t> storage(size);
   src.copyToMemBlock(storage);
-  bytes_ = storage.release();
+  setBytes(storage.release());
   table.incRefCount(statName());
 }
 
+SymbolTable::StoragePtr SymbolTableImpl::makeDynamicStorage(absl::string_view name) {
+  name = StringUtil::removeTrailingCharacters(name, '.');
+
+  // For all StatName objects, we first have the total number of bytes in the
+  // representation. But for inlined dynamic string StatName variants, we must
+  // store the length of the payload separately, so that if this token gets
+  // joined with others, we'll know much space it consumes until the next token.
+  // So the layout is
+  //   [ length-of-whole-StatName, LiteralStringIndicator, length-of-name, name ]
+  // So we need to figure out how many bytes we need to represent length-of-name and
+  // bytes
+
+  // payload_bytes is the total number of bytes needed to represent the characters
+  // in name, plus their encoded size, plus the literal indicator.
+  uint64_t payload_bytes = SymbolTableImpl::Encoding::totalSizeBytes(name.size()) + 1;
```

Comment:
nit: can you constify here and a bunch of other places in this PR where possible.